torch_geometric/explain/explainer.py

```diff
@@ -0,0 +1,342 @@
+
+        if self.type == "hard":
+            if self.value < 0 or self.value > 1:
+                raise ValueError(f'Invalid threshold value {self.value}. '
+                                 f'Valid values are in [0, 1].')
+        if self.type in ["topk", "connected", "topk_hard"]:
+            if self.value <= 0 or not isinstance(self.value, int):
+                raise ValueError(f'Invalid threshold value {self.value}. '
+                                 f'Valid values are positif integers.')
+
+
+@dataclass
+class ExplainerConfig:
+    """Class to store and validate high level explanation parameters."""
+    explanation_type: str
+    mask_type: str
+    _valid_explanation_type = ["model", "phenomenon"]
+    _valid_mask_type = [
+        "node", "edge", "node_and_edge", "layers", "node_feat", "edge_feat",
+        "node_feat_and_edge_feat", "node_feat_and_edge", "node_and_edge_feat"
+    ]
+
+    def __post_init__(self):
+        if self.explanation_type not in self._valid_explanation_type:
+            raise ValueError(
+                f'Invalid explanation type {self.explanation_type}. '
+                f'Valid types are {self._valid_explanation_type}.')
+        if self.mask_type not in self._valid_mask_type:
+            raise ValueError(f'Invalid mask type {self.mask_type}. '
+                             f'Valid types are {self._valid_mask_type}.')
+
+
+@dataclass
+class ModelConfig:
+    model: torch.nn.Module
+    return_type: str
+    task_level: str = "graph"
+    mode: str = "classification"
+    _valid_model_return_type = ["logits", "probs", "raw", "regression"]
+    _valid_model_mode = ["classification", "regression"]
+    _valid_task = ["graph", "node", "edge"]
+
+    def __post_init__(self):
+
+        if self.return_type not in self._valid_model_return_type:
+            raise ValueError(
+                f'Invalid model return type {self.return_type}. '
+                f'Valid types are {self._valid_model_return_type}.')
+        if self.mode not in self._valid_model_mode:
+            raise ValueError(f'Invalid model mode {self.mode}. '
+                             f'Valid modes are {self._valid_model_mode}.')
+
+        if self.task_level not in self._valid_task:
+            raise ValueError(f'Invalid task {self.task_level}. '
+                             f'Valid tasks are {self._valid_task}.')
+
+        if Data not in getfullargspec(self.model.forward).annotations.values():
+            message = "The model does not accept a `Data` object as input."
+            message += "Consider using `explain.utils.Interface` to wrap the"
+            message += "forward function of your model."
+            message += "Check the documentation for more details."
+
+            raise ValueError(message)
+
+        # update model mode based on return type
+        if self.return_type == "regression":
+            self.mode = "regression"
+
+
+class Explainer(torch.nn.Module):
```

Comment:
Sorry high level thought: Intuitively it would make more sense to me to have something like a builder rather than having the explainer be a module which wraps another module? Do we need both this abstraction and the `ExplainerAlgorithm`?

Reply:
I think you're right. I have to think about how explanations can be used in model training to check if the inheritance is needed  (like in _Schramowski, P., Stammer, W., Teso, S. et al. Making deep neural networks right for the right scientific reasons by interacting with their explanations. Nat Mach Intell 2, 476–486 (2020). https://doi.org/10.1038/s42256-020-0212-3_).

Reply:
For the post-processing of the explanation? (I know it's quite simple for now, but it may be upgraded).
This would ensure that the only responsibility of the `ExplainerAlgorithm` are to return masks and then the rest is always taken care of in the same way.

Reply:
I was thinking of `Explainer` as being a base class and the various `ExplainerAlgorithm` just implement the functions needed to provide the masks.

Similar to how our `Aggregation` package works.

Reply:
I'll have a look at the 'Aggregation' package then 👍

Reply:
Ok, I see what you meant. You're probably right since the `ExplainerAlgorithm` is only called once in `_compute_explanation`.  Maybe @RexYing has some more insights ?

Reply:
I don't have a strong opinion on this one, but I personally like the separation between algorithm and an `Explainer` that takes care of the rest. Makes it easy to see the concrete interface one has to implement for a new `ExplainerAlgorithm`.

Reply:
I think I will stay with the split then, I removed the useless fact that `Explainer` was a module (thanks @rusty1s  for the review btw 👍🏽 ), making it more clear.